src/Stratis.Bitcoin.Features.Consensus/Rules/ProvenHeaderRules/ProvenHeaderCoinstakeRule.cs

```diff
@@ -0,0 +1,251 @@
+    /// </list>
+    /// </summary>
+    /// <remarks>(*) - denotes rules that are expensive to execute.</remarks>
+    /// <seealso cref="T:Stratis.Bitcoin.Features.Consensus.Rules.ProvenHeaderRules.ProvenHeaderRuleBase" />
+    public class ProvenHeaderCoinstakeRule : ProvenHeaderRuleBase
+    {
+        /// <summary>The stake validator.</summary>
+        private IStakeValidator stakeValidator;
+
+        /// <inheritdoc />
+        public override void Initialize()
+        {
+            base.Initialize();
+
+            Guard.NotNull(this.PosParent.StakeValidator, nameof(this.PosParent.StakeValidator));
+            Guard.NotNull(this.PosParent.UtxoSet, nameof(this.PosParent.UtxoSet));
+
+            this.stakeValidator = this.PosParent.StakeValidator;
+        }
+
+        /// <inheritdoc/>
+        public override void Run(RuleContext context)
+        {
+            Guard.NotNull(context.ValidationContext.ChainedHeaderToValidate, nameof(context.ValidationContext.ChainedHeaderToValidate));
+
+            ChainedHeader chainedHeader = context.ValidationContext.ChainedHeaderToValidate;
+            int height = chainedHeader.Height;
+
+            if (context.SkipValidation || !this.IsProvenHeaderActivated(height))
+                return;
+
+            var header = (ProvenBlockHeader)chainedHeader.Header;
+
+            this.CheckCoinstakeIsNotNull(header);
+
+            FetchCoinsResponse coins = this.GetAndValidateCoins(header);
+            UnspentOutputs prevUtxo = this.GetAndValidatePreviousUtxo(coins);
+
+            this.CheckIfCoinstakeIsTrue(header);
+
+            this.CheckHeaderAndCoinstakeTimes(header);
+
+            this.CheckCoinstakeAgeRequirement(chainedHeader, prevUtxo);
+
+            this.CheckSignature(header, prevUtxo);
+
+            this.CheckStakeKernelHash((PosRuleContext)context, prevUtxo, header, chainedHeader);
+
+            this.CheckCoinstakeMerkleProof(header);
+
+            this.CheckHeaderSignatureWithConinstakeKernel(header, prevUtxo);
+        }
+
+        /// <summary>
+        /// Checks the coinstake to make sure it is not null.
+        /// </summary>
+        /// <param name="header">The proven blockheader.</param>
+        private void CheckCoinstakeIsNotNull(ProvenBlockHeader header)
+        {
+            if (header.Coinstake != null)
+                return;
+
+            this.Logger.LogTrace("(-)[COINSTAKE_IS_NULL]");
+            ConsensusErrors.EmptyCoinstake.Throw();
+        }
+
+        /// <summary>
+        /// Fetches and validates coins from coins view.
+        /// </summary>
+        /// <param name="header">The header.</param>
+        private FetchCoinsResponse GetAndValidateCoins(ProvenBlockHeader header)
+        {
+            // First try finding the previous transaction in database.
+            TxIn txIn = header.Coinstake.Inputs[0];
+            FetchCoinsResponse coins = this.PosParent.UtxoSet.FetchCoinsAsync(new[] { txIn.PrevOut.Hash }).GetAwaiter().GetResult();
```

Comment:
`txIn.PrevOut.Hash` should be available in CoinSet (is this a FV rule?)

Reply:
This is the point of this rule to check that this is available in Proven headers.  Proven header rules run before the FV rules

Reply:
I am assuming PH download will go far beyond maturity so a coin may not be present anymore in coinview if it was spend already.